ray-curator/ray_curator/models/nsfw.py

```diff
@@ -0,0 +1,165 @@
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Apply normalization to input tensor.
+
+        Args:
+            x: Input tensor to normalize.
+
+        Returns:
+            Normalized tensor.
+        """
+        return (x - self.mean) / self.variance.sqrt()
+
+
+class NSFWModel(nn.Module):
+    """NSFW detection model.
+
+    A neural network that processes CLIP embeddings to predict NSFW scores.
+    Based on LAION's CLIP-based-NSFW-Detector.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the NSFW model.
+
+        Args:
+            None
+        """
+        super().__init__()
+        self.norm = Normalization([768])
+        self.linear_1 = nn.Linear(768, 64)
+        self.linear_2 = nn.Linear(64, 512)
+        self.linear_3 = nn.Linear(512, 256)
+        self.linear_4 = nn.Linear(256, 1)
+        self.act = nn.ReLU()
+        self.act_out = nn.Sigmoid()
+
+    @torch.no_grad()
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass through the NSFW model.
+
+        Args:
+            x: Input embeddings tensor.
+
+        Returns:
+            NSFW probability scores.
+        """
+        x = self.norm(x)
+        x = self.act(self.linear_1(x))
+        x = self.act(self.linear_2(x))
+        x = self.act(self.linear_3(x))
+        return self.act_out(self.linear_4(x))  # type: ignore[no-any-return]
+
+
+class NSFWScorer(ModelInterface):
+    """Public interface for NSFW scoring of image embeddings.
+
+    This class provides a standardized interface for scoring the likelihood
+    of images containing sexually explicit material using a pre-trained model.
+    """
+
+    def __init__(self, model_dir: str) -> None:
+        """Initialize the NSFW scorer interface."""
+        super().__init__()
+        self.device = "cuda:0" if torch.cuda.is_available() else "cpu"
+        self.dtype = torch.float32
+        self.model_dir = model_dir
+        # These will be initialized in setup()
+        self.nsfw_model = None
+        self.weights_path = None
+
+    @property
+    def conda_env_name(self) -> str:
+        """Get the name of the conda environment required for this model.
+
+        Returns:
+            Name of the conda environment.
+        """
+        return "video_splitting"
+
+    @property
+    def model_id_names(self) -> list[str]:
+        """Get the model ID names associated with this NSFW scorer.
+
+        Returns:
+            A list containing the model ID for NSFW scoring.
+        """
+        return [_NSFW_MODEL_ID]
+
+    def setup(self) -> None:
+        """Set up the NSFW scoring model by loading weights."""
+        weights_filename = "clip_autokeras_binary_nsfw.pth"
+        self.weights_path = str(Path(self.model_dir) / self.model_id_names[0] / weights_filename)
+
+        # Download weights if they don't exist
+        if not os.path.exists(self.weights_path):
+            self._download_weights()
```

Comment:
This should be `setup_on_node`, not `setup`. Then can be called from the `ProcessingStage`'s `setup_on_node`.